cranelift/codegen/meta/src/gen_listing.rs

```diff
@@ -0,0 +1,171 @@
+//! Generate human-readable lists of instructions and their encodings.
+//!
+//! To display the generated contents, one could use
+//! `find target -name 'listing-x86*' | xargs cat | less` from the top-level project directory.
+use std::collections::BTreeMap;
+use std::fmt::{Debug, Display, Error, Formatter as DisplayFormatter};
+
+use crate::cdsl::encodings::EncodingContent;
+use crate::cdsl::isa::TargetIsa;
+use crate::cdsl::recipes::{EncodingRecipe, EncodingRecipeNumber, OperandConstraint};
+use crate::cdsl::regs::{IsaRegs, RegClass, RegClassIndex};
+use crate::error;
+use crate::shared::Definitions as SharedDefinitions;
+use crate::srcgen::Formatter;
+
+/// Emit a list of all encodings in a human-readable format.
+fn emit_listings(_: &SharedDefinitions, isa: &TargetIsa, fmt: &mut Formatter) {
+    let mut recipe_counter = BTreeMap::new();
+    for cpu_mode in &isa.cpu_modes {
+        fmtln!(fmt, "CPU Mode: {}", cpu_mode.name);
+        fmtln!(fmt, "==================");
+
+        // Sort encodings alphabetically.
+        let mut sorted_encodings = cpu_mode.encodings.clone();
+        sorted_encodings.sort_by(|a, b| a.inst().name.cmp(&b.inst().name));
+
+        // Print each encoding.
+        for encoding in &sorted_encodings {
+            let printable = DisplayEncoding::new(encoding, isa);
+            fmtln!(fmt, "{}", printable);
+            let counter = recipe_counter.entry(encoding.recipe).or_insert(0);
+            *counter += 1;
+        }
+        fmtln!(fmt, "");
+    }
+
+    // Print usage counts of used recipes.
+    fmtln!(fmt, "Used Recipes");
+    fmtln!(fmt, "==================");
+
+    // Sort used recipes by most frequently used.
+    let mut sorted_recipes: Vec<_> = recipe_counter.iter().collect();
+    sorted_recipes.sort_by(|a, b| b.1.cmp(a.1));
+
+    for (recipe, count) in sorted_recipes {
+        let recipe = DisplayRecipe::new(*recipe, isa);
+        fmtln!(fmt, "{} uses of recipe({})", count, recipe);
+    }
+    fmtln!(fmt, "");
+}
+
+/// Helper structure for pretty-printing encodings.
+struct DisplayEncoding<'a> {
+    encoding: &'a EncodingContent,
+    isa: &'a TargetIsa,
+}
+
+impl<'a> DisplayEncoding<'a> {
+    fn new(encoding: &'a EncodingContent, isa: &'a TargetIsa) -> Self {
+        Self { encoding, isa }
+    }
+
+    fn recipe(&self) -> &EncodingRecipe {
```

Comment:
Since this is used only once in the `fmt` impl below, can you inline it there, please?